### ISSUE TITLE
allow adding PIDs for workflows content types

### ISSUE THREAD

**Issue by laureD19** [SSHOC/sshoc-marketplace-backend#445]
based on DARIAH PIDs

**Comment by laureD19** [comment]
during the creation/publication of workflow, make a call to the Handle server and ask for a PID
Add the PID from the Handle to `externalID` attribute, with the "Handle" type.

Contact point to get the endpoint and the credentials: @vronk or @dpancic

**Comment by pWoz** [comment]
Good morning. I'm about to start to work on this issue. Can you (@vronk or @dpancic) give me some more information about that handle server? (endpoint, credentials, api)

**Comment by KlausIllmayer** [comment]
Something to be discussed with @vronk - I don't think we have agreed on a specific PID server.

**Comment by stefanprobst** [comment]
@KlausIllmayer we have a prefix for sshoc marketplace, see https://redmine.acdh.oeaw.ac.at/projects/dsie/wiki/Handle

**Comment by KlausIllmayer** [comment]
Thanks for pointing me to this issue (it is an internal link) - but I guess we still need to discuss with @vronk how to setup it for marketplace.

**Comment by vronk** [comment]
Handle for MP is:    21.11160 

endpoint is: vm02.pid.gwdg.de
    Port: 2645/8004

Credentials we will convey to you separately. Or is it key-based authentication, @dpancic 

@pWoz, let us know, if you need any more details or guidance.

If so, @dpancic can probably give best answers.

**Comment by dpancic** [comment]
Exactly, for the SSH Market Place Prefix 300:0.NA/21.11160, we received the key for Handle API key-based authentication from GWDG.

Here is the documentation for Handle API key-based authentication: [Handle API Documentation](https://docs.pidinst.org/en/latest/epic-cookbook/handles.html#using-the-handle-api)

**Comment by vronk** [comment]
@pWoz, were you able to look at this again?

**Comment by pWoz** [comment]
Not yet @vronk . But I'm going to do it in the nearest days.

**Comment by vronk** [comment]
Thank you. 
We would be grateful, if this could be prioritized.

**Comment by pWoz** [comment]
I have several questions related with the implementation:

1. As I understand suffix should be generated automatically by the handle server. In our case we should use `VersionedItem.persistentId`, right? So handle service will do the redirection to (for example) this workflow: https://marketplace.sshopencloud.eu/workflow/P0siWJ. Is that correct?
2. I'm not pretty sure if I understand the communication with handle server correctly. So correct me if I'm wrong:

- there seems to be two APIs that can be used, described here: https://docs.pidinst.org/en/latest/epic-cookbook/handles.html#generating-a-new-instrument-pid
`ePIC API` and `Handle API`. Which one should I use? Actually I don't see big difference between them.

- basically while creating new Workflow we should do the request (assuming that we are using ePIC API) like the following:
`curl -v -u "username:password" -H "Accept:application/json" -H "Content-Type:application/json" -X POST --data '[{"type":"URL","parsed_data":"https://marketplace.sshopencloud.eu/workflow/P0siWJ"}]' http://vm04.pid.gwdg.de:8081/handles/21.11160/`
In that correct?

- It seams that this service (listening on port 8081) is test service protected by username and password. Is it possible to get this credentials from somewhere for testing?

- which one is the production service? There seems to be multiple services listening on multiple ports 8001,8002,8003,8004. Can I use username and password for communication here or certificates as it is in examples for `Handle API`?

**Comment by vronk** [comment]
@dpancic, can you maybe comment on this? (or who could help?)

**Comment by vronk** [comment]
There is only  a certificate-based authentication currently for this prefix.

The curl call for key-based authentication is (in general):
`
curl -k -s --insecure ${VERBOSE} --key ${USERKEY} --cert ${USERCERT} -H "Content-Type:application/json" -H 'Authorization: Handle clientCert="true"' -X PUT --data  '{"values":[{"index":100,"type":"HS_ADMIN","data":{"value":{"index":'${INDEX}',"handle":"'${PREFIX}'\/'${USER}'","permissions":"011111110011","format":"admin"},"format":"admin"}},{"index":1,"type":"URL","data":"www.gwdg.de"}]}' https://${SERVPORT}/api/handles/${PREFIX}/TEST3
`

@KlausIllmayer, @stefanprobst, can you maybe help with this? 
Did you manage to communicate with the handle-server with key-based auth?

**Comment by stefanprobst** [comment]
> Did you manage to communicate with the handle-server with key-based auth?

yes.

although IIRC letting the handle server create suffixes via POST had some issues, but PUT and providing an identifier works.

this is what we do for dariah campus: https://github.com/DARIAH-ERIC/dariah-campus/blob/main/lib/server/handle/create-handle.ts

**Comment by pWoz** [comment]
Integration should work now.

1. This mechanism have to be configured in `application.yml `
2. Workflow persistent identifiers are used there so Handle Server will redirect to the workflow's home page (i.e. https://marketplace.sshopencloud.eu/workflow/P0siWJ)
3. There is no need to create External Source definition in the system beforehand. It will be automatically generated when needed;
4. In case of any issues with Handle server workflow will still be generated but without `externalId`.

Pull request is here:
https://github.com/SSHOC/sshoc-marketplace-backend/pull/499

**Comment by pWoz** [comment]
I've just started to modify Handle server integration so PIDs will point to the workflow versions (not just to the latest workflow) as we discussed it during our meeting. 

Where persistent identifiers (stored on Handle server) should point to?
In the current implementation it points to the workflow description web page (i.e. https://marketplace.sshopencloud.eu/workflow/P0siWJ)
but I cannot find web page that describes one specific workflow version. 

On the back-end we have the following endpoint: /api/workflows/ZLKn3K/versions/1104

**Comment by KlausIllmayer** [comment]
Access to a specific version of an workflow is only granted to specific users. To see on frontend the versions (and their urls) you must login, then a history-button will appear,where you can see all the workflow versions.
Take as example this workflow: https://marketplace.sshopencloud.eu/workflow/P0siWJ
You won't see a history, when you are an anonymous user. But when you login, you will see a history-button with all all workflow versions: https://marketplace.sshopencloud.eu/workflow/P0siWJ/history
The urls are quite the same as for the api-endpoints, e.g. https://marketplace.sshopencloud.eu/workflow/P0siWJ/version/60709
generalized: https://marketplace.sshopencloud.eu/workflow/[workflow-pid]/version/[workflow-id]
So I guess this are the urls that you could use to point to.

But access to these versions is currently only allowed to the user, who created the version and to users with a role of moderator or administrator. All others will get an access denied error. Maybe something to discuss in this issues: https://github.com/SSHOC/sshoc-marketplace-backend/issues/501 @mkrzmr

**Comment by pWoz** [comment]
Thanks for your clarification. I modified the code, handles points now to the workflow versions:
https://marketplace.sshopencloud.eu/workflow/[workflow-pid]/version/[workflow-id]

Solution is ready to be revied.

**Comment by KlausIllmayer** [comment]
I was able to create on stage a workflow using the parameter `createHandle=true` which produced a working handle PID. It points to the version of the workflow, which means, it is currently only accessible with the given URL when you are logged into the SSHOMP. We need to relax who can show versions of a workflow. Will create an issue on this.
